### ISSUE TITLE
wish: lazy-load libraries

### ISSUE THREAD

**Issue by markstos** [dolox/fallback#10]
One nice feature of require.js is that it supports "lazy loading"
libraries. That is, it can avoid loading libraries that are not needed.

Here's my own use-case for it, and how I think it might be designed and
implemented here.

We use both jQuery and jQuery-UI on our site, and jQuery UI is large
enough that we would prefer not to load it on pages where we don't need
it.

With the current design, I could address this by creating two small
scripts:

```
# Set up definition to load jQuery with fall back
load-jquery.js

# Set up definition to load jQuery with fall back
load-jquery-and-jquery-ui.js
```

I would then load whichever one I needed, and find-and-replace my jQuery
"ready" commands to this instead:

```
fallback.ready(function() {
    // Completed
});
```

The way I would rather use fallback.js to put all my CDN and fallback
references in one file, and then I would just declare which ones I
needed on a given page, and just those would be loaded.

Here's how that might look:

```
# My local script which calls fallback.load()
cdn-fallbacks.js
```

Then, I would replace the jQuery ready() calls depending on which ones I
wanted to load:

```
# Load everything by default (current behavior)
fallback.ready(function() {
    // Completed
});

# explicitly load just jQuery
fallback.ready(['jQuery'],function() {
    // Completed
});

# explicitly load just jQuery and jQuery-UI
fallback.ready(['jQuery','jQuery-UI'],function() {
    // Completed
});
```

In the implementation, the call to "this.initialize();" would be moved
from "load()" to "ready()" and would take as argument the optional array
that can now be passed to ready(). If the list has entries in it, only
those elements would be loaded. The call to initialize(); run
immediately when ready() is called (after the callbacks are setup).

require.js supports basically the same feature with the same syntax, and
I assume it's implemented in a similar way.

Without this feature, I could end up with a proliferation of files
related to fallback.js, that each load different combinations of
modules. Or more likely, I might consider switchng to using require.js
rather than deal with the lack of missing feature.

Thanks for your consideration.

```
Mark
```

**Comment by sgarbesi** [comment]
If you need something more in depth an alternative library is your best bet.

I'm not against adding more features as long as they make sense and don't cause the library to get large enough to where it's deemed inadequate for it's sole purpose. One of the nice features about this library is its size, it's extremely fast for a user to load it, which is really the only reason you would want to use this over RequireJS or many of the other libraries out there.

Now to point out the feature you requested. Maybe the comment I'm about to make is more from a engineering standpoint than feature standpoint. After you load a library such as jQuery or jQuery UI from a CDN (majority of cases) the CDN will cache the file on the client side. Therefore the next go around the user isn't actually downloading it again per say, it's a one time deal.

I've dealt with coding on high traffic sites, for something like those extremely important libraries that the majority of the site uses it makes sense to load them everywhere, as after they are loaded once they are cached and not downloaded again by the client.

Also maybe I'm misunderstanding part of your example, but if you're not including jQuery UI on a page and you're executing jQuery UI code, won't you end up with JavaScript errors? As long as the code you're using isn't executing the library all should be well with or without the feature in question?

Also if you wanted to something like this now:

```
fallback.ready(['jQuery'],function() {
    // Completed
});
```

...you actually already can. What's stopping you from doing this to achieve the functionality you desire?

```
fallback.ready(function() {
    if (window.jQuery) {
        // Launch jQuery Dependent Ready Function
    }
});
```

You'll have to sell me on this one.

Again like I said I'm not against adding features, but the main benefit is the library file size. This is what separates this from RequireJS. Sure RequireJS has a ton of features packed in, but the question you have to ask yourself is this. Do you really need all that additional overhead? Features you may never even use or tap into? These are questions I cannot answer. For myself in particular, I don't and won't, the main reason I wrote this library from the door.

**Comment by markstos** [comment]
Thanks for the reply.

Some of my pages only use jQuery and not jQuery-UI. jQuery-UI is almost 200K, and I want to avoid sending to the browser when I can. I realize some people will already have it cached from Google's CDN, but some won't have the specific version we are using. Adding a unnecessary 200K to a page load is not the kind of first impression I want to give visitors who come to our site.

In the code example you provide, both jQuery and jQuery-UI would be loaded, assuming they are both defined elsewhere. I only want to load jQuery-UI if I'm actually going to use it. 

Regarding library size, the difference between fallback.js and require.js is only about 10k.  Size is important to me too, which I'm why I'm concerned about not loading a 200K jQuery-UI library when I don't need to. I suspect it would be a bigger performance win for me to switch to require.js so that I can more easily avoid loading big libraries that aren't used on page, then it is save a few kilobytes off of the size of my loader library. 

As I sketched out the feature, I don't think it would add much "weight" to the library. It would probably be about moving the initialization from being up-front to being on-demand, and only loading the requested libraries.

**Comment by sgarbesi** [comment]
Doesn't using a `if statement` within the `ready` function provide the functionality?

The code would have to:

```
1. attach each ready function to an array of objects with the libraries that are specified
2. loop through every library in the array of loaded libraries every time a library has finished loading followed by another loop through the ready functions object arrays to see if the properly libraries were loaded and whether or not it's time for invocation.
```

Not hard to do, but does it really beat just using a `if statement`?

RequireJS minified from their website is 14.65 KB, Fallback JS is 1.69 KB, aka 8x (round up, or 9x if you want to round properly) difference.

I'm actually going to remove the `ready_invoke` variable I added that gives you the ability to set it to false so that the callbacks aren't fired off, I feel like it's a waste of size in the code.

Aside from that I actually sketched up a splash page and am going to grab a domain for the site. I appreciate the advice and feedback.
